js/web/lib/wasm/jsep/webgpu/ops/quantize-linear.ts

```diff
@@ -0,0 +1,219 @@
+      const outputSize = ShapeUtil.size(outputShape);
+      const isPacked = inputType === DataType.int8 || inputType === DataType.uint8;
+      const inputShape = isPacked ? [Math.ceil(ShapeUtil.size(inputs[0].dims) / 4)] : inputs[0].dims;
+      const scaleShape = inputs[1].dims;
+      const zeroPointInput = inputs.length > 2 ? inputs[2] : undefined;
+      const zeroPointShape = zeroPointInput ?
+          (isPacked ? [Math.ceil(ShapeUtil.size(zeroPointInput.dims) / 4)] : zeroPointInput.dims) :
+          undefined;
+      // Scales input is a scaler for per-tensor/per-layer quantization, 1-D tensor for per-axis quantization
+      // or tensor with same rank as input for blocked quantization.
+      const perLayerQuantization = scaleShape.length === 0 || (scaleShape.length === 1 && scaleShape[0] === 1);
+      const perAxisQuantization = perLayerQuantization === false && scaleShape.length === 1;
+      // Left unnecessary commented-out assignment for documentation
+      // const blockQuantization = perLayerQuantization === false && perAxisQuantization === false;
+      const maxComponents = getMaxComponents(outputSize);
+      const useComponents = perLayerQuantization && (!isPacked || maxComponents === 4);
+      const components = useComponents ? maxComponents : 1;
+      const inputComponent = (useComponents && !isPacked) ? maxComponents : 1;
+      const input = inputVariable('input', isPacked ? DataType.uint32 : inputType, inputShape.length, inputComponent);
+      const scale = inputVariable('scale', dataType, scaleShape.length);
+      const zeroPoint = zeroPointInput ?
+          inputVariable('zero_point', isPacked ? DataType.uint32 : inputType, zeroPointShape!.length) :
+          undefined;
+      const output = outputVariable('output', dataType, outputShape.length, components);
+      const inputVariables = [input, scale];
+      if (zeroPoint) {
+        inputVariables.push(zeroPoint);
+      }
+      const inputShapes = [inputShape, scaleShape];
+      if (zeroPointInput) {
+        inputShapes.push(zeroPointShape!);
+      }
+      const programUniforms: ProgramUniform[] = [
+        {type: DataType.uint32, data: outputSize / components}, {type: DataType.uint32, data: axis},
+        {type: DataType.uint32, data: attributes.blockSize}, ...createTensorShapeVariables(...inputShapes, outputShape)
+      ];
+      const getShaderSource = (shaderHelper: ShaderHelper) => {
+        const uniforms: UniformsArrayType =
+            [{name: 'output_size', type: 'u32'}, {name: 'axis', type: 'u32'}, {name: 'block_size', type: 'u32'}];
+        return `
+      ${shaderHelper.registerUniforms(uniforms).declareVariables(...inputVariables, output)}
+      ${shaderHelper.mainStart()}
+          ${shaderHelper.guardAgainstOutOfBoundsWorkgroupSizes('uniforms.output_size')}
+          let output_indices = ${output.offsetToIndices('global_idx')};
+
+          // Set input x
+          ${(() => {
```

Comment:
`${isPacked ? `xxx` : `yyy` }` is simpler.

Reply:
Because we have multiple lines code, using if is more readable than using ternary operator
```
          ${isPacked ?
            `
            let input = ${input.getByOffset('global_idx / 4')};
            let x_vec = ${isSigned ? 'unpack4xI8(input)' : 'unpack4xU8(input)'};
            let x_value = ${components === 1 ? 'x_vec[global_idx % 4]' : 'x_vec'};` :
            `let x_value = ${input.getByOffset('global_idx')};`
          }
```